bidict/_base.py

```diff
@@ -26,30 +26,31 @@
 #==============================================================================
 
 
-"""Provides :class:`BidictBase`."""
+"""Provide :class:`BidictBase`."""
 
+import typing as _t
 from collections import namedtuple
-from collections.abc import Mapping
 from copy import copy
 from weakref import ref
 
 from ._abc import BidirectionalMapping
-from ._dup import ON_DUP_DEFAULT, RAISE, DROP_OLD, DROP_NEW
-from ._exc import (
-    DuplicationError, KeyDuplicationError, ValueDuplicationError, KeyAndValueDuplicationError)
-from ._sntl import _MISS, _NOOP
+from ._dup import ON_DUP_DEFAULT, RAISE, DROP_OLD, DROP_NEW, OnDup
+from ._exc import DuplicationError, KeyDuplicationError, ValueDuplicationError, KeyAndValueDuplicationError
+from ._typing import _NONE, KT, VT, OKT, OVT, MapOrIterItems
 from ._util import _iteritems_args_kw
 
 
-_DedupResult = namedtuple('_DedupResult', 'isdupkey isdupval invbyval fwdbykey')
 _WriteResult = namedtuple('_WriteResult', 'key val oldkey oldval')
-_NODUP = _DedupResult(False, False, _MISS, _MISS)
+_DedupResult = namedtuple('_DedupResult', 'isdupkey isdupval invbyval fwdbykey')
+_NODUP = _DedupResult(False, False, _NONE, _NONE)
+
+BT = _t.TypeVar('BT', bound='BidictBase')  # typevar for BidictBase.copy
 
 
-class BidictBase(BidirectionalMapping):
+class BidictBase(BidirectionalMapping[KT, VT]):
     """Base class implementing :class:`BidirectionalMapping`."""
 
-    __slots__ = ('_fwdm', '_invm', '_inv', '_invweak', '_hash', '__weakref__')
+    __slots__ = ['_fwdm', '_invm', '_inv', '_invweak', '_hash', '__weakref__']
```

Comment:
unrelated?

Reply:
No, not unrelated. If you read the rest of the changes in this file, you'll see this is for the https://bugs.python.org/issue41451 workaround below (removing the `'__weakref__'` slot) after this class definition.